src/caliper/Wrapped.h

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2015, Lawrence Livermore National Security, LLC.  
+// Produced at the Lawrence Livermore National Laboratory.
+//
+// This file is part of Caliper.
+// Written by David Boehme, boehme3@llnl.gov.
+// LLNL-CODE-678900
+// All rights reserved.
+//
+// For details, see https://github.com/scalability-llnl/Caliper.
+// Please also see the LICENSE file for our additional BSD notice.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+//  * Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the disclaimer below.
+//  * Redistributions in binary form must reproduce the above copyright notice, this list of
+//    conditions and the disclaimer (as noted below) in the documentation and/or other materials
+//    provided with the distribution.
+//  * Neither the name of the LLNS/LLNL nor the names of its contributors may be used to endorse
+//    or promote products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+// OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// LAWRENCE LIVERMORE NATIONAL SECURITY, LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/// \file AnnotatedRegion.h
+/// Caliper C++ Function Annotation Library
+
+#ifndef CALI_ANNOTATED_REGION_H
+#define CALI_ANNOTATED_REGION_H
+
+#include "Annotation.h"
+#include <type_traits>
+
+namespace cali{
+
+template<typename LB, typename... Args>
+auto Wrapped(LB body, const char* name, Args... args) -> typename std::result_of<LB(Args...)>::type{
+    cali::Annotation::Guard func_annot(cali::Annotation("Wrapped Function").begin(name));
+    return body(args...);
+}
+
+template<typename LB, typename... Args>
+auto Wrapped(const char* name, LB body, Args... args) -> typename std::result_of<LB(Args...)>::type{
+    cali::Annotation::Guard func_annot(cali::Annotation("Wrapped Function").begin(name));
```

Comment:
It's better to make that Annotation object static, otherwise you'll do a string lookup on every invocation. Maybe also consider a different name for the attribute; Caliper should handle arbitrary names now but external processing might be easier if it doesn't have spaces.